c++/src/kj/compat/http.h

```diff
@@ -769,21 +769,47 @@ struct HttpServerSettings {
   // completes, we'll let the connection stay open to handle more requests.
 };
 
-class HttpServer final: private kj::TaskSet::ErrorHandler {
+class HttpServerErrorHandler {
+public:
+  virtual kj::Promise<void> requestError(
+      uint statusCode, kj::StringPtr statusText, kj::String errorMessage,
+      HttpHeaderTable& headerTable, kj::Maybe<HttpService::Response&> response) = 0;
+  // Similar to `HttpService::request()`, but called when the HttpServer encounters any sort of
+  // error, be it a protocol error or an exception thrown by the HttpService. The parameters
+  // `statusCode`, `statusText`, and `errorMessage`, are hints from the HttpServer suggesting what
+  // sort of response it thinks we should send to the client. The default implementation (which you
+  // must explicitly opt into by calling `HttpServerErrorHandler::requestError()` from your derived
+  // class) sends just such a response.
+  //
+  // If the HttpService did not already send a partial response, the `response` parameter will be
+  // non-null and available for you to customize the response sent to the client. You could use this
+  // to perform error message sanitization, e.g. to censor stack traces.
+  //
+  // Unlike `HttpService::request()`, when calling `response.send()` in this context, a
+  // "Connection: close" header will be added, and the connection will be closed.
+  //
+  // TODO(cleanup): It might make more sense to pass in the original exception object. But this
+  //   would not handle the 400 Bad Request case, which is the original motivation for this whole
+  //   exercise.
+};
+
+class HttpServer final: private kj::TaskSet::ErrorHandler, private HttpServerErrorHandler {
   // Class which listens for requests on ports or connections and sends them to an HttpService.
 
 public:
   typedef HttpServerSettings Settings;
   typedef kj::Function<kj::Own<HttpService>(kj::AsyncIoStream&)> HttpServiceFactory;
 
   HttpServer(kj::Timer& timer, HttpHeaderTable& requestHeaderTable, HttpService& service,
-             Settings settings = Settings());
+             Settings settings = Settings(),
+             kj::Maybe<HttpServerErrorHandler&> errorHandler = nullptr);
   // Set up an HttpServer that directs incoming connections to the given service. The service
   // may be a host service or a proxy service depending on whether you are intending to implement
   // an HTTP server or an HTTP proxy.
 
   HttpServer(kj::Timer& timer, HttpHeaderTable& requestHeaderTable,
-             HttpServiceFactory serviceFactory, Settings settings = Settings());
+             HttpServiceFactory serviceFactory, Settings settings = Settings(),
+             kj::Maybe<HttpServerErrorHandler&> errorHandler = nullptr);
```

Comment:
Can `errorHandler` be a member of the `settings` struct?

Reply:
Yes, but it will be source-incompatible because the settings objects can no longer be literals. I'm fine with that, but didn't play with it enough to know what all needs to be changed. I'll find out.

Reply:
Not sure what you mean. What kind of existing code would break, exactly?

Reply:
`constexpr HttpServerSettings LOCAL_CLIENT_SETTINGS = { ... };`

Probably not a big deal -- the place we'd need to modify in Workers we'd need to modify to add the new setting anyway, I just didn't spend any time seeing where else that pattern is used. I agree it'd be preferable as a setting field.

Reply:
(I'm on the bus right now, so not really explaining well.)

Reply:
Is this just because `Maybe<T&>`'s default constructor is not `constexpr`? Could we make it `constexpr`?